src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedCatchParameterShouldBeUnnamedCheck.java

```diff
@@ -0,0 +1,270 @@
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Ensures that catch parameters that are not used are declared as an unnamed variable.
+ * </p>
+ * <p>
+ * Rationale:
+ * </p>
+ * <ul>
+ *     <li>
+ *         Improves code readability by clearly indicating which parameters are unused.
+ *     </li>
+ *     <li>
+ *         Follows Java conventions for denoting unused parameters with an underscore ({@code _}).
+ *     </li>
+ * </ul>
+ * <p>
+ * See the <a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html">
+ * Java Language Specification</a> for more information about unnamed variables.
+ * </p>
+ * <p>
+ * <b>Attention</b>: This check should be activated only on source code
+ * that is compiled by jdk21 or higher;
+ * unnamed catch parameters came out as the first preview in Java 21.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code unused.catch.parameter}
+ * </li>
+ * </ul>
+ *
+ * @since 10.18.0
+ *
+ */
+
+@FileStatefulCheck
+public class UnusedCatchParameterShouldBeUnnamedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_CATCH_PARAMETER = "unused.catch.parameter";
+
+    /**
+     * Invalid parents of the catch parameter identifier.
+     */
+    private static final int[] INVALID_CATCH_PARAM_IDENT_PARENTS = {
+        TokenTypes.DOT,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.TYPE,
+    };
+
+    /**
+     * Keeps track of the catch parameters in a block.
+     */
+    private final Deque<CatchParameterDetails> catchParameters = new ArrayDeque<>();
```

Comment:
Yes but we should be sure that we will never leave any date behind from file to file; [as we always pop the top of the stack ](https://github.com/mahfouz72/checkstyle/blob/827ccf0a7da00b12d3480027540569203c389555/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedCatchParameterShouldBeUnnamedCheck.java#L137)when we leave the catch token. So, when we finish the file (leave all tokens) by then we should have an empty stack.

Does it make sense to clear the collection even if we are sure that it will always be cleared using other ways (`pop()`s in this example

Reply:
> Does it make sense to clear the collection even if we are sure that it will always be cleared using other ways (pop()s in this example

But you can't be sure. You are only sure your check processes good granted everything external doesn't fail, however you can't guarantee external things won't fail. If another check has an NPE, and halt on exception is turned off, it will leave that file without calling anymore leave tokens and start the next file with data left on the stack.

Reply:
Yea, got it. Done

Reply:
See `isStatefulFieldClearedDuringBeginTree` to applease the pitest gods.

Reply:
done